### PR TITLE
[SPARK-36292][SQL] Refactor third set of 20 in QueryExecutionErrors to use error classes

### PR DESCRIPTION
**Pull request by dgd-contributor**
### What changes were proposed in this pull request?
Refactor some exceptions in QueryExecutionErrors to use error classes.
```
cannotGenerateCodeForUnsupportedTypeError
cannotInterpolateClassIntoCodeBlockError
customCollectionClsNotResolvedError
classUnsupportedByMapObjectsError
nullAsMapKeyNotAllowedError
methodNotDeclaredError
constructorNotFoundError
primaryConstructorNotFoundError
unsupportedNaturalJoinTypeError
notExpectedUnresolvedEncoderError
unsupportedEncoderError
notOverrideExpectedMethodsError
failToConvertValueToJsonError
unexpectedOperatorInCorrelatedSubquery
unreachableError
unsupportedRoundingMode
resolveCannotHandleNestedSchema
inputExternalRowCannotBeNullError
fieldCannotBeNullMsg
fieldCannotBeNullError
```

### Why are the changes needed?
There are currently ~350 exceptions in this file; so this PR only focuses on the third set of 20.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existed UT Testcase

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @karenfeng FYI

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -3,9 +3,25 @@
...
+    "message" : [ "Cannot generate code for unsupported type: %s" ],
+    "sqlState" : "42000"
+  },
+  "CANNOT_INTERPOLATE_CLASS_INTO_CODE_BLOCK" : {
+    "message" : [ "Can not interpolate %s into code block." ],
+    "sqlState" : "42000"
+  },
+  "CLASS_UNSUPPORTED_BY_MAP_OBJECTS" : {
+    "message" : [ "Class `%s` is not supported by `MapObjects` as resulting collection." ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:16]: 0A000 may be a better fit here.

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -21,6 +37,14 @@
   "FAILED_SET_ORIGINAL_PERMISSION_BACK" : {
     "message" : [ "Failed to set original permission %s back to the created path: %s. Exception: %s" ]
   },
+  "FAIL_TO_CONVERT_VALUE_TO_JSON" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:40]: For consistency: FAIL -> FAILED?

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -77,29 +110,73 @@
...
+  "PRIMARY_CONSTRUCTOR_NOT_FOUND" : {
+    "message" : [ "Couldn't find a primary constructor on %s" ],
+    "sqlState" : "42000"
+  },
   "RENAME_SRC_PATH_NOT_FOUND" : {
     "message" : [ "Failed to rename as %s was not found" ],
     "sqlState" : "22023"
   },
+  "RESOLVE_CANNOT_HANDLE_NESTED_SCHEMA" : {
+    "message" : [ "Can not handle nested schema yet...  plan %s" ],
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:138]: We should [avoid](https://spark.apache.org/error-message-guidelines.html) making promises about future support. Can you remove `yet`?

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -77,29 +110,73 @@
...
+  },
   "SECOND_FUNCTION_ARGUMENT_NOT_INTEGER" : {
     "message" : [ "The second argument of '%s' function needs to be an integer." ],
     "sqlState" : "22023"
   },
   "UNABLE_TO_ACQUIRE_MEMORY" : {
     "message" : [ "Unable to acquire %s bytes of memory, got %s" ]
   },
+  "UNEXPECTED_OPERATOR_IN_CORRELATED_SUBQUERY" : {
+    "message" : [ "Unexpected operator %s in correlated subquery%s" ],
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:149]: Nit: there should be a space between `subquery` and `%s`

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -77,29 +110,73 @@
...
+    "sqlState" : "42000"
+  },
+  "UNREACHABLE" : {
+    "message" : [ "This line should be unreachable%s" ],
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:153]: Nit: there should be a space between `unreachable` and `%s`

File: core/src/main/resources/error/error-classes.json [hunk 6]
```diff
@@ -77,29 +110,73 @@
...
+    "sqlState" : "42000"
+  },
   "UNRECOGNIZED_SQL_TYPE" : {
     "message" : [ "Unrecognized SQL type %s" ],
     "sqlState" : "42000"
   },
+  "UNSUPPORTED_ENCODER" : {
+    "message" : [ "Only expression encoders are supported for now." ],
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:161]: We should [avoid](https://spark.apache.org/error-message-guidelines.html) making promises about future support. Can you remove `for now`?

File: core/src/main/resources/error/error-classes.json [hunk 7]
```diff
@@ -77,29 +110,73 @@
...
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:162]: 0A000 may be a better fit for all "unsupported" errors.

File: core/src/main/resources/error/error-classes.json [hunk 8]
```diff
@@ -77,29 +110,73 @@
...
+  },
   "UNSUPPORTED_LITERAL_TYPE" : {
     "message" : [ "Unsupported literal type %s %s" ],
     "sqlState" : "0A000"
   },
+  "UNSUPPORTED_NATURAL_JOIN_TYPE" : {
+    "message" : [ "Unsupported natural join type %s" ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:170]: 0A000 may be a better fit for all "unsupported" errors.

File: core/src/main/scala/org/apache/spark/SparkException.scala [hunk 9]
```diff
@@ -224,3 +224,69 @@ class SparkSQLFeatureNotSupportedException(errorClass: String, messageParameters
   override def getErrorClass: String = errorClass
   override def getSqlState: String = SparkThrowableHelper.getSqlState(errorClass)
 }
+
+/**
+ * Unsupported operation exception thrown from Spark with an error class.
+ */
+class SparkUnsupportedOperationException(errorClass: String, messageParameters: Array[String])
```
Comment (karenfeng) [inline core/src/main/scala/org/apache/spark/SparkException.scala:231]: We should make these classes package-private, eg. https://github.com/apache/spark/blob/f78d8394dcf19891141e353ea3b6a76020faf844/core/src/main/scala/org/apache/spark/SparkException.scala#L234

**Review by karenfeng — COMMENTED** [review]
Thanks for working on this @dgd-contributor! I left some comments.

File: core/src/main/resources/error/error-classes.json [hunk 10]
```diff
@@ -55,7 +83,12 @@
     "sqlState" : "22023"
   },
   "INVALID_JSON_SCHEMA_MAPTYPE" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:85]: While we're here, can you also update: `INVALID_JSON_SCHEMA_MAPTYPE` -> `INVALID_JSON_MAP_KEY_TYPE`? The original class name is not very descriptive.

File: core/src/main/resources/error/error-classes.json [hunk 11]
```diff
@@ -55,7 +83,12 @@
...
-    "message" : [ "Input schema %s can only contain StringType as a key type for a MapType." ]
+    "message" : [ "Input schema %s can only contain StringType as a key type for a MapType." ],
+    "sqlState" : "42000"
+  },
+  "METHOD_NOT_DECLARED" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:89]: I don't think this is used? It's also already defined as `MISSING_METHOD`.

File: core/src/main/resources/error/error-classes.json [hunk 12]
```diff
@@ -77,29 +110,73 @@
     "message" : [ "PARTITION clause cannot contain a non-partition column name: %s" ],
     "sqlState" : "42000"
   },
+  "NOT_EXPECTED_UNRESOLVED_ENCODER" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:113]: `NOT_EXPECTED_UNRESOLVED_ENCODER` -> `UNEXPECTED_UNRESOLVED_ENCODER`

File: core/src/main/resources/error/error-classes.json [hunk 13]
```diff
@@ -77,29 +110,73 @@
...
+    "message" : [ "Unresolved encoder expected, but %s was found." ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:115]: This looks like an internal error, let's remove the SQLSTATE.

File: core/src/main/resources/error/error-classes.json [hunk 14]
```diff
@@ -77,29 +110,73 @@
...
+  },
+  "NOT_OVERRIDE_EXPECTED_METHODS" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:117]: `NOT_OVERRIDE_EXPECTED_METHODS` -> `EXPECTED_METHODS_NOT_OVERRIDDEN`

File: core/src/main/resources/error/error-classes.json [hunk 15]
```diff
@@ -77,29 +110,73 @@
...
+    "message" : [ "%s must override either %s or %s" ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:119]: This looks like an internal error, let's remove the SQLSTATE.

File: core/src/main/resources/error/error-classes.json [hunk 16]
```diff
@@ -77,29 +110,73 @@
...
+    "message" : [ "Cannot use null as map key!" ],
+    "sqlState" : "42000"
+  },
   "PIVOT_VALUE_DATA_TYPE_MISMATCH" : {
     "message" : [ "Invalid pivot value '%s': value data type %s does not match pivot column data type %s" ],
     "sqlState" : "42000"
   },
+  "PRIMARY_CONSTRUCTOR_NOT_FOUND" : {
+    "message" : [ "Couldn't find a primary constructor on %s" ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:131]: This looks like an internal error, let's remove the SQLSTATE.

File: core/src/main/resources/error/error-classes.json [hunk 17]
```diff
@@ -77,29 +110,73 @@
...
+  },
   "RENAME_SRC_PATH_NOT_FOUND" : {
     "message" : [ "Failed to rename as %s was not found" ],
     "sqlState" : "22023"
   },
+  "RESOLVE_CANNOT_HANDLE_NESTED_SCHEMA" : {
+    "message" : [ "Can not handle nested schema...  plan %s" ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:139]: This looks like an internal error, let's remove the SQLSTATE.

File: core/src/main/resources/error/error-classes.json [hunk 18]
```diff
@@ -77,29 +110,73 @@
...
   },
+  "PRIMARY_CONSTRUCTOR_NOT_FOUND" : {
+    "message" : [ "Couldn't find a primary constructor on %s" ],
+    "sqlState" : "42000"
+  },
   "RENAME_SRC_PATH_NOT_FOUND" : {
     "message" : [ "Failed to rename as %s was not found" ],
     "sqlState" : "22023"
   },
+  "RESOLVE_CANNOT_HANDLE_NESTED_SCHEMA" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:137]: `RESOLVE_CANNOT_HANDLE_NESTED_SCHEMA` -> `CANNOT_RESOLVE_NESTED_SCHEMA`

File: core/src/main/resources/error/error-classes.json [hunk 19]
```diff
@@ -77,29 +110,73 @@
...
   "UNABLE_TO_ACQUIRE_MEMORY" : {
     "message" : [ "Unable to acquire %s bytes of memory, got %s" ]
   },
+  "UNEXPECTED_OPERATOR_IN_CORRELATED_SUBQUERY" : {
+    "message" : [ "Unexpected operator %s in correlated subquery %s" ],
+    "sqlState" : "42000"
+  },
+  "UNREACHABLE" : {
+    "message" : [ "This line should be unreachable %s" ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:154]: This looks like an internal error, let's remove the SQLSTATE.

File: core/src/main/resources/error/error-classes.json [hunk 20]
```diff
@@ -77,29 +110,73 @@
...
     "sqlState" : "22023"
   },
   "UNABLE_TO_ACQUIRE_MEMORY" : {
     "message" : [ "Unable to acquire %s bytes of memory, got %s" ]
   },
+  "UNEXPECTED_OPERATOR_IN_CORRELATED_SUBQUERY" : {
+    "message" : [ "Unexpected operator %s in correlated subquery %s" ],
+    "sqlState" : "42000"
+  },
+  "UNREACHABLE" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:152]: `UNREACHABLE` -> `REACHED_UNREACHABLE_LINE` - we should provide more detail

File: core/src/main/resources/error/error-classes.json [hunk 21]
```diff
@@ -77,29 +110,73 @@
...
+  },
   "UNSUPPORTED_LITERAL_TYPE" : {
     "message" : [ "Unsupported literal type %s %s" ],
     "sqlState" : "0A000"
   },
+  "UNSUPPORTED_NATURAL_JOIN_TYPE" : {
+    "message" : [ "Unsupported natural join type %s" ],
+    "sqlState" : "0A000"
+  },
+  "UNSUPPORTED_OPERATION_NOT_RESOLVED" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:172]: `UNSUPPORTED_OPERATION_NOT_RESOLVED` -> `NOT_RESOLVED`

**Review by karenfeng — COMMENTED** [review]
Thanks for working on this @dgd-contributor! I left some comments, please tag me when you address them.

**Comment by dgd-contributor** [comment]
cc @karenfeng, can you pls review it again? 
Thank you.

File: core/src/main/resources/error/error-classes.json [hunk 22]
```diff
@@ -17,6 +17,14 @@
   "CANNOT_GENERATE_CODE_FOR_EXPRESSION" : {
     "message" : [ "Cannot generate code for expression: %s" ]
   },
+  "CANNOT_GENERATE_CODE_FOR_UNSUPPORTED_TYPE" : {
+    "message" : [ "Cannot generate code for unsupported type: %s" ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:22]: This should be 0A000 (unsupported)

File: core/src/main/resources/error/error-classes.json (hunk 16 above)
Comment (karenfeng) [reply core/src/main/resources/error/error-classes.json:170]: Bump!

File: core/src/main/resources/error/error-classes.json (hunk 18 above)
Comment (karenfeng) [reply core/src/main/resources/error/error-classes.json:176]: Bump

File: core/src/main/resources/error/error-classes.json [hunk 23]
```diff
@@ -149,10 +206,22 @@
...
+  },
   "UNSUPPORTED_LITERAL_TYPE" : {
     "message" : [ "Unsupported literal type %s %s" ],
     "sqlState" : "0A000"
   },
+  "UNSUPPORTED_NATURAL_JOIN_TYPE" : {
+    "message" : [ "Unsupported natural join type %s" ],
+    "sqlState" : "0A000"
+  },
+  "UNSUPPORTED_ROUNDING" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:221]: UNSUPPORTED_ROUNDING -> UNSUPPORTED_ROUNDING_MODE

File: core/src/main/resources/error/error-classes.json [hunk 24]
```diff
@@ -137,6 +187,13 @@
   "UNABLE_TO_ACQUIRE_MEMORY" : {
     "message" : [ "Unable to acquire %s bytes of memory, got %s" ]
   },
+  "UNEXPECTED_OPERATOR_IN_CORRELATED_SUBQUERY" : {
+    "message" : [ "Unexpected operator %s in correlated subquery %s" ],
+    "sqlState" : "42000"
+  },
+  "UNEXPECTED_UNRESOLVED_ENCODER" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:194]: Whoops, I meant: UNEXPECTED_UNRESOLVED_ENCODER -> EXPECTED_UNRESOLVED_ENCODER

File: core/src/main/resources/error/error-classes.json [hunk 25]
```diff
@@ -118,14 +150,32 @@
     "message" : [ "PARTITION clause cannot contain a non-partition column name: %s" ],
     "sqlState" : "42000"
   },
+  "NOT_RESOLVED" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:153]: In retrospect, this is specific to custom collection class resolution. Can we make this class name more specific? Eg. `CUSTOM_COLLECTION_CLASS_NOT_RESOLVED`

File: core/src/main/resources/error/error-classes.json [hunk 26]
```diff
@@ -49,6 +68,14 @@
   "FAILED_SET_ORIGINAL_PERMISSION_BACK" : {
     "message" : [ "Failed to set original permission %s back to the created path: %s. Exception: %s" ]
   },
+  "FAILED_TO_CONVERT_VALUE_TO_JSON" : {
+    "message" : [ "Failed to convert value %s (class of %s) with the type of %s to JSON." ],
+    "sqlState" : "22005"
+  },
+  "FIELD_CANNOT_BE_NULL" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:75]: Let's make this more specific: `ROW_FIELD_CANNOT_BE_NULL`

**Review by karenfeng — COMMENTED** [review]
I left a few more comments. Thanks for working on this @dgd-contributor, please ping me when you've addressed them.

**Comment by dchvn** [comment]
Thank you for your review, @karenfeng. As we and @HyukjinKwon discussed contributing by sharing accounts, we will close this PR, and recreate this one with an individual account. We will also update the code with your comments.

**Comment by dgd-contributor** [comment]
Thank you for your review, @karenfeng. As we and @HyukjinKwon discussed contributing by sharing accounts, we will close this PR, and recreate this one with an individual account. We will also update the code with your comments.

**Pull request closed without merging**